clippy_lints/src/if_let_mutex.rs

```diff
@@ -0,0 +1,139 @@
+use crate::utils::{match_type, paths, span_lint_and_help};
+use if_chain::if_chain;
+use rustc_hir::{Arm, Expr, ExprKind, MatchSource, Stmt, StmtKind};
+use rustc_lint::{LateContext, LateLintPass};
+use rustc_session::{declare_lint_pass, declare_tool_lint};
+
+declare_clippy_lint! {
+    /// **What it does:** Checks for `Mutex::lock` calls in `if let` expression
+    /// with lock calls in any of the else blocks.
+    ///
+    /// **Why is this bad?** The Mutex lock remains held for the whole
+    /// `if let ... else` block and deadlocks.
+    ///
+    /// **Known problems:** None.
+    ///
+    /// **Example:**
+    ///
+    /// ```rust,ignore
+    /// if let Ok(thing) = mutex.lock() {
+    ///     do_thing();
+    /// } else {
+    ///     mutex.lock();
+    /// }
+    /// ```
+    /// Should be written
+    /// ```rust,ignore
+    /// let locked = mutex.lock();
+    /// if let Ok(thing) = locked {
+    ///     do_thing(thing);
+    /// } else {
+    ///     use_locked(locked);
+    /// }
+    /// ```
+    pub IF_LET_MUTEX,
+    correctness,
+    "locking a `Mutex` in an `if let` block can cause deadlocks"
+}
+
+declare_lint_pass!(IfLetMutex => [IF_LET_MUTEX]);
+
+impl LateLintPass<'_, '_> for IfLetMutex {
+    fn check_expr(&mut self, cx: &LateContext<'_, '_>, ex: &'_ Expr<'_>) {
+        if_chain! {
+            if let ExprKind::Match(ref op, ref arms, MatchSource::IfLetDesugar {
+                contains_else_clause: true,
+            }) = ex.kind; // if let ... {} else {}
+            if let ExprKind::MethodCall(_, _, ref args) = op.kind;
+            let ty = cx.tables.expr_ty(&args[0]);
+            if match_type(cx, ty, &paths::MUTEX); // make sure receiver is Mutex
+            if method_chain_names(op, 10).iter().any(|s| s == "lock"); // and lock is called
```

Comment:
I don't like, that the `10` is hardcoded to check method chains (and that only method chains are checked. I'd suggest to implement a visitor here, that visits the `op`, searches for a `mutex.lock()` call, and then returns the `mutex`. After this, you could implement an `ExprUseVisitor`, that searches for this `mutex` expression. If the `parent_expr(mutex)` is then a `MethodCall` to `lock`, the lint should be triggered.

An `ExprUseVisitor` can be implemented similar to 
https://github.com/rust-lang/rust-clippy/blob/e004b9431d608bc6fcceb7f175138dccce163eb5/clippy_lints/src/utils/usage.rs#L66-L78

For this, it should be enough to implement the `borrow(..)` method, since the `lock(&self)` function borrows the `mutex`.

Reply:
If you need help/further instructions on this, let me know.